### PR TITLE
Allow creating DataArrays with nD coordinate variables

### LINKED ISSUE #8106: Plotting accessor error for datasets with Dimension name matches coordinate names
### What happened?

I am working with NetCDF files produced from the [Model Evaluation Tools (MET)](https://[dtcenter.org](https://dtcenter.org/community-code/model-evaluation-tools-met)/community-code/model-evaluation-tools-met).

This feature in xarray v2023.8.0 

> allows reading of datasets where a dimension name is associated with a multidimensional variable

lets me read the file where the dimension names `lat` and `lon` match the coordinate names `lat` `lon` (see https://github.com/dtcenter/MET/issues/1246). This is wonderful! Thank you for this 🎉🙌🏻

```python
import urllib
import xarray as xr

# download a sample MET file
urllib.request.urlretrieve(
    "https://github.com/dtcenter/MET/raw/main_v11.1/data/poly/NCEP_masks/WEST_mask.nc",
    "WEST_mask.nc",
)

# open MET file
ds = xr.open_dataset(
    "WEST_mask.nc",
    engine="netcdf4",
)
ds
```
image

However, one issue I've run into is that the plotting accessor doesn't seem to work for this case

```python
ds.WEST.plot()

OUT:
ValueError: coordinate 'lat' is a DataArray dimension, but it has shape (110, 147) rather than expected shape 110 matching the dimension size
```

To overcome this, I had to rename the dims

```python
ds.rename_dims({"lat": "y", "lon": "x"}).WEST.plot()
```
image

### What did you expect to happen?

I expected the plot accessor to work without needing to rename the dimensions.

### Minimal Complete Verifiable Example

```Python
import urllib
import xarray as xr

# download a sample MET file
urllib.request.urlretrieve(
    "https://github.com/dtcenter/MET/raw/main_v11.1/data/poly/NCEP_masks/WEST_mask.nc",
    "WEST_mask.nc",
)

ds = xr.open_dataset(
    "WEST_mask.nc",
    engine="netcdf4",
)
ds.WEST.plot()
```

### MVCE confirmation

- [X] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [X] Complete example — the example is self-contained, including all data and the text of any traceback.
- [X] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [X] New issue — a search of GitHub Issues suggests this is not a duplicate.

### Relevant log output

_No response_

### Anything else we need to know?

_No response_

### Environment

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.11.4 | packaged by conda-forge | (main, Jun 10 2023, 18:08:17) [GCC 12.2.0]
python-bits: 64
OS: Linux
OS-release: 3.10.0-1160.53.1.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.12.2
libnetcdf: 4.9.1

xarray: 2023.8.0
pandas: 2.0.3
numpy: 1.25.2
scipy: 1.11.2
netCDF4: 1.6.3
pydap: None
h5netcdf: 1.2.0
h5py: 3.8.0
Nio: None
zarr: None
cftime: 1.6.2
nc_time_axis: None
PseudoNetCDF: None
iris: None
bottleneck: None
dask: 2023.8.1
distributed: 2023.8.1
matplotlib: 3.7.2
cartopy: 0.22.0
seaborn: 0.12.2
numbagg: None
fsspec: 2023.6.0
cupy: None
pint: 0.22
sparse: None
flox: None
numpy_groupies: None
setuptools: 68.1.2
pip: 23.2.1
conda: None
pytest: 7.4.0
mypy: None
IPython: 8.7.0
sphinx: 4.5.0
/p/home/blaylock/miniconda3/envs/flight/lib/python3.11/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")

</details>

### PR DESCRIPTION
**Pull request by dcherian**
- [x] Closes #8106
- [x] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

cc @blaylockbk

### CODE REVIEW

File: xarray/tests/test_dataarray.py [hunk 1]
```diff
@@ -7075,3 +7072,32 @@ def test_error_on_ellipsis_without_list(self) -> None:
         da = DataArray([[1, 2], [1, 2]], dims=("x", "y"))
         with pytest.raises(ValueError):
             da.stack(flat=...)  # type: ignore
+
+
+def test_nD_coord_dataarray():
```
Comment (headtr1ck) [inline xarray/tests/test_dataarray.py:7077]:
```suggestion
def test_nD_coord_dataarray() -> None:
```

File: xarray/tests/test_dataarray.py [hunk 2]
```diff
@@ -7075,3 +7072,32 @@ def test_error_on_ellipsis_without_list(self) -> None:
...
+
+    expected = da.drop_vars("x")
+    _, actual = xr.broadcast(da, da2)
+    assert_identical(expected, actual)
+
+    actual, _ = xr.broadcast(da, da3)
+    expected = da.expand_dims(z=4, axis=-1)
+    assert_identical(actual, expected)
+
+    da2 = DataArray(np.ones((2, 4)), coords={"x": 0}, dims=["x", "y"])
```
Comment (headtr1ck) [inline xarray/tests/test_dataarray.py:7101]:
```suggestion
    da4 = DataArray(np.ones((2, 4)), coords={"x": 0}, dims=["x", "y"])
```

File: xarray/tests/test_dataarray.py [hunk 3]
```diff
@@ -7075,3 +7072,32 @@ def test_error_on_ellipsis_without_list(self) -> None:
...
+    assert "x" not in da2.xindexes
```
Comment (headtr1ck) [inline xarray/tests/test_dataarray.py:7102]:
```suggestion
    assert "x" not in da4.xindexes
```

File: xarray/tests/test_dataarray.py [hunk 4]
```diff
@@ -7075,3 +7072,32 @@ def test_error_on_ellipsis_without_list(self) -> None:
...
+    assert "x" in da2.coords
```
Comment (headtr1ck) [inline xarray/tests/test_dataarray.py:7103]:
```suggestion
    assert "x" in da4.coords
```

**Review by headtr1ck — APPROVED** [review]
Looks good

File: xarray/tests/test_dataarray.py [hunk 5]
```diff
@@ -7085,3 +7082,32 @@ def test_error_on_ellipsis_without_list(self) -> None:
         da = DataArray([[1, 2], [1, 2]], dims=("x", "y"))
         with pytest.raises(ValueError):
             da.stack(flat=...)  # type: ignore
+
+
+def test_nD_coord_dataarray() -> None:
```
Comment (dcherian) [inline xarray/tests/test_dataarray.py:7087]: @benbovy can you review these tests please

Comment (dcherian) [reply xarray/tests/test_dataarray.py:7115]: gentle ping @benbovy

**Review by benbovy — APPROVED** [review]
LGTM too, thanks @dcherian

**Pull request merged**